cupy/random/_bit_generator.pyx

```diff
@@ -0,0 +1,183 @@
+# distutils: language = c++
+import threading
+
+import numpy
+
+from libc.stdint cimport intptr_t, uint64_t, uint32_t
+
+import cupy
+from cupy.cuda cimport stream
+from cupy.core.core cimport ndarray
+from cupy.random._generator import init_curand, random_raw
+
+# We need access to the sizes here, so this is why we have this header
+# in here instead of cupy backends
+cdef extern from 'cupy_distributions.cuh' nogil:
+    cppclass curandState:
+        pass
+    cppclass curandStateMRG32k3a:
+        pass
+    cppclass curandStatePhilox4_32_10_t:
+        pass
+
+    cdef enum _RandGenerators 'RandGenerators':
+        CURAND_XOR_WOW
+        CURAND_MRG32k3a
+        CURAND_PHILOX_4x32_10
+
+
+class BitGenerator:
+    """Generic BitGenerator.
+
+    Base Class for generic BitGenerators, which provide a stream
+    of random bits based on different algorithms. Must be overridden.
+
+    Args:
+        seed (int, array_like[ints], numpy.random.SeedSequence, optional):
+            A seed to initialize the `BitGenerator`. If None, then fresh,
+            unpredictable entropy will be pulled from the OS. If an ``int`` or
+            ``array_like[ints]`` is passed, then it will be passed to
+            ~`numpy.random.SeedSequence` to derive the initial `BitGenerator`
+            state. One may also pass in a `SeedSequence` instance.
+    """
+    def __init__(self, seed=None):
+        self.lock = threading.Lock()
+        # TODO(ecastill) port SeedSequence
+        if isinstance(seed, numpy.random.SeedSequence):
+            self._seed_seq = seed
+        else:
+            if isinstance(seed, cupy.ndarray):
+                seed = cupy.asnumpy(seed)
+            self._seed_seq = numpy.random.SeedSequence(seed)
+        dev = cupy.cuda.Device()
```

Comment:
How about using `cupy.cuda.get_device_id` to avoid instantiating `Device` object?

Reply:
Good catch! thanks